core/src/reservation.rs

```diff
@@ -27,6 +30,47 @@ use scale_info::{
     TypeInfo,
 };
 
+/// An unchangeable wrapper over u64 value, which is required
+/// to be used as a "view-only" reservations nonce in a message
+/// execution context.
+///
+/// ### Note:
+/// By contract, It must be instantiated only once, when message execution
+/// context is created. Also the latter is required to be instantiated only
+/// once, when incoming dispatch is created.
+#[derive(
+    Clone, Copy, Default, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo,
+)]
+pub struct ReservationNonce(u64);
+
+impl From<&InnerNonce> for ReservationNonce {
+    fn from(nonce: &InnerNonce) -> Self {
+        ReservationNonce(nonce.0)
+    }
+}
+
+/// A changeable wrapper over u64 value, which is required
+/// to be used as an "active" reservations nonce in a gas reserver.
+#[derive(Debug, Clone)]
+struct InnerNonce(u64);
+
+impl InnerNonce {
+    /// Fetches current state of the nonce and
+    /// updates its state by incrementing it.
+    fn fetch_inc_nonce(&mut self) -> u64 {
```

Comment:
`nonce.fetch_inc_nonce()` seems strange for me
```suggestion
    fn fetch_inc(&mut self) -> u64 {
```

Reply:
done